doc/dump_smart.1
.TH DUMP_SMART 1 "2010 Feb 9" "ASN.1 Tools" "ROA-PKI"

.SH NAME
dump_smart \- converts ASN.1-encoded RPKI file to a more readable form

.SH SYNOPSIS
\fBdump_smart\fP \fIfilename\fP

.SH DESCRIPTION
This program translates the contents of the file
.I filename
into more readable form and outputs the result to stdout.  The type of
file is determined by the suffix in
.IR filename ", which must be one of"
.BR .cer " for a certificate, " .crl " for a CRL, " .man
.RB " for a manifest, " .roa " for a ROA, or " .rta " for an RPKI TA."

.PP
If there is no error, the translated contents of the file are printed
out on stdout in the readable form.  If there is an error, one of
these messages is printed on stderr:

.IP \(bu 4
Usage: name of input file
.IP \(bu 4
Suffix is missing
.IP \(bu 4
Unknown type ...
.IP \(bu 4
Error reading at ...

.PP
where the ellipsis is replaced with specific information.  For the
"Unknown type" error it is the suffix.  For the "Error reading at"
error it is a sequence of dot-separated numbers indicating where the
error was found, e.g. 1.2.3.4 means in the fourth member of the third
member of the second member of the first member of the ASN.1 encoding.

.SH ALTERNATIVES

.B dump_smart
outputs a raw human-readable format which still contains enough
information to be edited and converted directly back into a
certificate using ASN.1 tools.  An even more human-friendly format
(but NOT easily reversible) can be obtained by issuing the openssl
command:
.IP
openssl x509 -inform DER -text < \fIfilename\fP

.SH AUTHOR
Charlie Gardiner.  Documentation edited by Andrew Chi.

.SH "SEE ALSO"
.BR dump (1),
.BR make_oidtable (1)


ChangeLog
next

	* Support Ghostbusters Records [RFC 6493].

0.7

	* Add preliminary support for adding files from the local cache. In the
	  future, this could be used to quickly deploy additional relying party
	  machines without each one needing to download all files from the
	  global RPKI.
	* Fix a buffer overflow when manifests list filenames longer than 199
	  bytes.
	* Fix a read of uninitialized memory when the first file listed on a
	  manifest has an unknown extension.
	* Don't allow empty sets of IP resources in certificates.
	  (http://tools.ietf.org/html/rfc6487#section-4.8.10)
	* Don't allow Subsequent Address Family Identifier (SAFI) values in IP
	  resources. (http://tools.ietf.org/html/rfc6487#section-4.8.10)
	* Fix bugs in the ROA validation code where AS numbers equal to zero
	  or larger than 2^31 - 1 were rejected.

0.6

	* Simplify the commands used during normal operation. The commands
	  are now named after what the user intends to do, instead of after the
	  internal component we used to implement them. Some of the scripts are
	  reorganized so that each script does one function (e.g., synchronize
	  with remote repositories) instead of presenting an interface to one
	  internal component (e.g., the component that validates objects fed to
	  it by other components while syncronizing with remote repositories).
	* Use a single configuration file syntax instead of a mix of
	  environment variables and multiple incompatible configuration files.
	  Additionally, provide an example configuration file with comments
	  describing each option. This should make rpstir easier to set up.
	* Switch from one Makefile per directory to a single Makefile. Aside
	  from speeding up parallel compilation, this lets us simplify
	  internal dependencies. With one Makefile per directory, we had to
	  carefully order the directories such that nothing in one directory
	  depended on something in a later directory. Since the dependencies
	  changed over time, this became more and more complex to maintain.
	  With a single Makefile, we can just tell "make" about each dependency
	  and let "make" figure out the order to build things.
	* Remove duplicate code. Before switching to a single Makefile (see
	  above), we occasionally copied functions from one file to another to
	  avoid having to restructure the source tree and reorder the
	  per-directory Makefiles. With a single Makefile keeping track of
	  dependencies, we consolidated the duplicated code and let "make"
	  do the build ordering for us.
	* Find and remove unused code.
	* Improve support for CHECKTOOL. When $CHECKTOOL is set to "valgrind"
	  for the self-tests, much more code is now tested with valgrind.
	* Make it possible to install to the system instead of running from
	  within the source directory.
	* When an rsync download of a trust anchor fails, retry a few times
	  instead of giving up on downloading that trust anchor.
	* Support Cryptlib 3.4.2.
	* Fix a bug that occured when a trust anchor was added after its
	  children. Some of the children that should have been considered valid
	  were considered unknown. This doesn't happen in normal operation, but
	  was discovered during testing at IETF 85.
	* Don't require access to the database root account. As long as
	  rpstir's database user has full permissions on a database, there's
	  no reason to require access to the root database user.

0.5

	* Fix compatibility with older rsync servers.
	* Be more careful about runaway processes in a few places.
	* Log when a URI is dropped from the list to synchronize.
	* Minimize extraneous fetching of remote repositories.
	* Support 20-octet serial numbers, manifest numbers, and CRL numbers.
	* Check for new versions and log when a newer version is available.

0.4

	* Add conformance cases for various specifications and fix problems revealed by those conformance cases.
	* Performance tuning for rsync.
	* Remove dependency on python-netaddr and increase performance where python-netaddr used to be used.
	* Fix build and tests on NetBSD.
	* Fix compiling/linking with pthreads on some platforms.
	* More tests can be run under valgrind.
	* Various bug and compiler warning fixes.

0.3

	* Fix security bugs in rsync_cord and rsync_listener.
	* First-pass at preventing SQL injection attacks.
	* Update chaser to prevent shell injection attacks and to minimize duplicate downloads.
	* Warn if people try to build on an (unsupported) 64-bit platform or if OpenSSL was compiled without RFC 3779 support.
	* Add error checking to rsync_aur.
	* Performance tuning for rpki-rtr.
	* Various bug fixes.

0.2

	* Support current version of rpki-rtr protocol.
	* Various bug fixes.

0.1

	* Initial public release.


cg/tools/dump_smart.c

#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include "casn.h"
#include "roa.h"
#include "crlv2.h"

static char *msgs[] = {
    "Usage: name of input file\n",
    "Suffix missing in %s\n",
    "Unknown type %s\n",
    "Error reading at %s\n",
};

void fatal(
    int err,
    char *param)
{
    fprintf(stderr, msgs[err], param);
    if (err)
        exit(err);
}

int main(
    int argc,
    char **argv)
{
    struct ROA roa;
    struct Certificate certificate;
    struct CertificateRevocationList crl;
    char *buf;
    int bsize;


    if (argc < 2)
        fatal(0, (char *)0);
    char *p = strrchr(argv[1], (int)'.');
    if (!p)
        fatal(1, (char *)0);
    if (!strcmp(p, ".cer"))
    {
        Certificate(&certificate, (ushort) 0);
        if (get_casn_file(&certificate.self, argv[1], 0) < 0)
            fatal(3, casn_err_struct.asn_map_string);
        bsize = dump_size(&certificate.self);
        buf = (char *)calloc(1, bsize + 8);
        dump_casn(&certificate.self, buf);
        printf(buf);
        free(buf);
        delete_casn(&certificate.self);
    }
    else if (!strcmp(p, ".crl"))
    {
        CertificateRevocationList(&crl, (ushort) 0);
        if (get_casn_file(&crl.self, argv[1], 0) < 0)
            fatal(3, casn_err_struct.asn_map_string);
        bsize = dump_size(&crl.self);
        buf = (char *)calloc(1, bsize + 8);
        dump_casn(&crl.self, buf);
        printf(buf);
        free(buf);
        delete_casn(&crl.self);
    }
    else if (!strcmp(p, ".man") || !strcmp(p, ".mft") || !strcmp(p, ".mnf") ||
             !strcmp(p, ".roa") || !strcmp(p, ".rta"))
    {
        ROA(&roa, (ushort) 0);
        if (get_casn_file(&roa.self, argv[1], 0) < 0)
            fatal(3, casn_err_struct.asn_map_string);
        bsize = dump_size(&roa.self);
        buf = (char *)calloc(1, bsize + 8);
        dump_casn(&roa.self, buf);
        printf(buf);
        free(buf);
        delete_casn(&roa.self);
    }
    else
        fatal(2, p);
    return 0;
}
